### PR TITLE
[KYUUBI #4745] Support Flink's LocalZonedTimestamp DataType

### PR DESCRIPTION
**Pull request by link3280**
### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [X] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala [hunk 1]
```diff
@@ -739,6 +739,23 @@ abstract class FlinkOperationSuite extends HiveJDBCTestHelper with WithFlinkTest
...
 
+  test("execute statement - select timestamp with local time zone") {
+    withJdbcStatement() { statement =>
+      statement.executeQuery("CREATE VIEW T1 AS SELECT TO_TIMESTAMP_LTZ(4001, 3)")
+      statement.executeQuery("SET 'table.local-time-zone' = 'UTC'")
+      val resultSetUTC = statement.executeQuery("SELECT * FROM T1")
+      val metaData = resultSetUTC.getMetaData
+      assert(metaData.getColumnType(1) === java.sql.Types.OTHER)
+      assert(resultSetUTC.next())
+      assert(resultSetUTC.getString(1) === "1970-01-01 00:00:04.001 UTC")
```
Comment (link3280) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/operation/FlinkOperationSuite.scala:750]: `TIMESTAMP WITH LOCAL ZONE` should not show the zone part. Something is wrong.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala [hunk 2]
```diff
@@ -377,4 +398,23 @@ object RowSet {
         other.toString
     }
   }
+
+  /** should stay in sync with [[org.apache.hadoop.hive.common.type.TimestampTZUtil]] */
+  var TIMESTAMP_LZT_FORMATTER: DateTimeFormatter = {
+    val builder = new DateTimeFormatterBuilder()
+    // Date part
+    builder.append(DateTimeFormatter.ofPattern("uuuu-MM-dd"))
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala:406]: what does `uuuu-MM-dd` mean?

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala [hunk 3]
```diff
@@ -377,4 +398,23 @@ object RowSet {
...
+    // Time part
+    builder.optionalStart().appendLiteral(" ").append(
+      DateTimeFormatter.ofPattern("HH:mm:ss")).optionalStart().appendFraction(
+      ChronoField.NANO_OF_SECOND,
+      1,
+      9,
+      true).optionalEnd().optionalEnd()
+    // Zone part
+    builder.optionalStart().appendLiteral(" ").optionalEnd()
+    builder.optionalStart().appendZoneOrOffsetId().optionalEnd()
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala:416]: if it's not intended, why add this?

File: externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/result/ResultSetSuite.scala [hunk 4]
```diff
@@ -44,9 +46,9 @@ class ResultSetSuite extends KyuubiFunSuite {
       .data(rowsNew)
       .build
 
-    assert(RowSet.toRowBaseSet(rowsNew, resultSetNew)
-      === RowSet.toRowBaseSet(rowsOld, resultSetOld))
-    assert(RowSet.toColumnBasedSet(rowsNew, resultSetNew)
-      === RowSet.toColumnBasedSet(rowsOld, resultSetOld))
+    assert(RowSet.toRowBaseSet(rowsNew, resultSetNew, ZoneId.systemDefault())
```
Comment (pan3793) [inline externals/kyuubi-flink-sql-engine/src/test/scala/org/apache/kyuubi/engine/flink/result/ResultSetSuite.scala:49]: `ZoneId.systemDefault()` may cause different results in different environments, let's use a specific one, e.g. `America/Los_Angeles`(PST) `Asia/Shanghai`(CST)

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala (hunk 3 above)
Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala:416]: Because parsing at JDBC requires the zone part, otherwise we would ran into this exception: https://github.com/apache/kyuubi/blob/06dd7d3754ee210f7744e46030965a6f5df1953b/kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/common/TimestampTZUtil.java#L84

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala (hunk 2 above)
Comment (link3280) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/schema/RowSet.scala:406]: The format is copied from TimestampTZUtil in `kyuubi-hive-jdbc` module. IIUC, `uuuu` has better compatibility with years before `1970`.

**Comment by link3280** [comment]
@pan3793 I started a thread to discuss the timestamp format issue https://www.mail-archive.com/dev@flink.apache.org/msg66257.html. But I don't think it's a blocker for this PR.

**Review by pan3793 — APPROVED** [review]
Thanks @link3280, the implementation makes sense to me, and I agree the Flink different behavior should not block us to merge this one.

**Comment by pan3793** [comment]
Thanks, merged to master
